File: UnityScript/src/Scripts/OGTabs.js
#pragma strict

public class OGTabs extends OGWidget {
	public class Tab {
		public var title : String = "Tab";
		public var content : GameObject;
	}
	
	public var activeTab : int;
	public var tabs : List.<Tab> = new List.<Tab>();
	
	public function AddTab ( tabName : String, tabObject : GameObject, switchTo : boolean ) {
		var newTab : Tab = new Tab ();

		newTab.title = tabName;
		newTab.content = tabObject;

		tabs.Add ( newTab );

		if ( switchTo ) {
			activeTab = tabs.Count - 1;
		} else if ( tabs.Count < 2 ) {
			activeTab = 0;
		}
		
		Build ();
	}

	public function ClearTabs () {
		tabs.Clear ();
		activeTab = -1;
		Build ();
	}


	//////////////////
	// Set drawn
	//////////////////
	override function SetDrawn ( drawn : boolean ) {
		if ( tabs.Count != this.transform.childCount ) {
			Build ();
		}
		
		isDrawn = drawn;

		for ( var i : int = 0; i < this.transform.childCount; i++ ) {
			var btn : OGButton = this.transform.GetChild(i).GetComponent(OGButton);
			btn.SetDrawn ( isDrawn );
		}

		SetDirty ();
	}


	//////////////////
	// Build
	//////////////////
	override function Build () {
		isSelectable = true;

		var i : int = 0;
		var btn : OGButton;

		if ( !tabs ) {
			tabs = new List.<Tab>();
		}

		// Edit existing or create new ones
		for ( i = 0; i < tabs.Count; i++ ) {
			if ( i < this.transform.childCount ) {
				btn = this.transform.GetChild(i).GetComponent(OGButton);
			} else {
				btn = new GameObject ( "", OGButton ).GetComponent(OGButton);
				btn.transform.parent = this.transform;
			}

			btn.text = tabs[i].title;
			btn.hidden = true;
			btn.target = this.gameObject;
			btn.message = "SetActiveTab";
			btn.argument = i.ToString();
			btn.styles.basic = this.styles.basic;
			btn.styles.active = this.styles.active;

			btn.gameObject.name = i + ": " + tabs[i].title;
			btn.transform.localScale = new Vector3 ( 1.0 / tabs.Count, 1, 1 );
			btn.transform.localPosition = new Vector3 ( (i*1.0) / tabs.Count, 0, 0 );
		}

		// Destroy remaining
		if ( this.transform.childCount > tabs.Count ) {
			for ( i = tabs.Count; i < this.transform.childCount; i++ ) {
				DestroyImmediate ( this.transform.GetChild(i).gameObject );
			}
		}

		// Set active tab
		SetActiveTab ( activeTab );
	}

	public function SetActiveTab ( n : String ) {
		SetActiveTab ( int.Parse ( n ) );
	}

	public function SetActiveTab ( tab : int ) {
		if ( tab < 0 || tab >= tabs.Count ) { return; }
		
		activeTab = tab;
	
		for ( var i : int = 0; i < tabs.Count; i++ ) {
			var btn : OGButton = this.transform.Find(i + ": " + tabs[i].title).GetComponent(OGButton);
			
			if ( activeTab != i ) {
				btn.styles.basic = styles.basic;
				
				if ( tabs[i].content != null && tabs[i].content.activeSelf ) {
					tabs[i].content.SetActive ( false );
				}

			} else {
				btn.styles.basic = styles.active;
				
				if ( tabs[i].content != null && !tabs[i].content.activeSelf ) {
					tabs[i].content.SetActive ( true );
				}
			}
		}

		SetDirty ();
	}
	
	override function UpdateWidget () {
		// Null check
		if ( tabs.Count != this.transform.childCount ) {
			Build ();
		}

		// Update data
		if ( activeTab >= tabs.Count && tabs.Count > 0 ) {
			SetActiveTab ( tabs.Count - 1 );
	 	} 
	}
}

File: OpenGUI/Assets/Plugins/OpenGUI/Scripts/OGTabs.cs
using UnityEngine;
using System.Collections;
using System.Collections.Generic;

public class OGTabs : OGWidget {
	[System.Serializable]
	public class Tab {
		public string title = "Tab";
		public GameObject content;
	
		public Tab ( string title, GameObject content ) {
			this.title = title;
			this.content = content;
		}
	}
	
	public int activeTab;
	public string leftArrow = "«";
	public string rightArrow = "»";
	public List<Tab> tabs = new List<Tab>();
	public GameObject target;
	public string message;
	public string argument;
	public bool passSelectedTab = false;

	private bool overflow = false;
	private int maxTabs = 0;
	private int focusTab = 0;
	private bool leftNavDown = false;
	private bool rightNavDown = false;


	//////////////////
	// Rects
	//////////////////
	private float GetTotalWidth () {
		float width = 0f;
		float widestTab = 0f;

		for ( int i = 0; i < tabs.Count; i++ ) {
			float newWidth = OGDrawHelper.GetLabelWidth ( tabs[i].title, GetTabStyle ( i ).text );
			width += newWidth;

			if ( newWidth > widestTab ) {
				widestTab = newWidth;
			}
		}

		if ( width > drawRct.width ) {
			maxTabs = (int) Mathf.Clamp ( Mathf.Floor ( drawRct.width / widestTab ), 1f, tabs.Count );
		
		} else {
			maxTabs = tabs.Count;

		}

		return width;
	}
	
	private Rect GetTabRect ( int i ) {
		float tabWidth;
		
		if ( !overflow ) {
	       		tabWidth = drawRct.width / tabs.Count;

		} else {
			float width = drawRct.width;
			
			if ( activeTab > 0 ) {
				width -= drawRct.height;
			}
			
			if ( activeTab < tabs.Count - 1 ) {
				width -= drawRct.height;
			}

			tabWidth = width / maxTabs;
		
		}
		
		Rect rect = new Rect ( drawRct.x + tabWidth * ( i - focusTab ), drawRct.y, tabWidth, drawRct.height );
		
		if ( activeTab > 0 && overflow ) {
			rect.x += drawRct.height;
		}

		return rect;
	}

	private Rect GetLeftRect () {
		return new Rect ( drawRct.x, drawRct.y, drawRct.height, drawRct.height );
	}
	
	private Rect GetRightRect () {
		return new Rect ( drawRct.xMax - drawRct.height, drawRct.y, drawRct.height, drawRct.height );
	}

	
	//////////////////
	// Styles
	//////////////////
	private OGStyle GetTabStyle ( int i ) {
		if ( i == activeTab ) {
			return styles.active;
		} else {
			return styles.basic;
		}
	}


	//////////////////
	// Management
	//////////////////
	public void AddTab ( string tabName, GameObject tabObject, bool switchTo ) {
		Tab newTab = new Tab ( tabName, tabObject );

		tabs.Add ( newTab );

		if ( switchTo ) {
			SetActiveTab ( tabs.Count - 1 );
		} else if ( tabs.Count < 2 ) {
			SetActiveTab ( 0 );
		}
	}

	public void ClearTabs () {
		tabs.Clear ();
		activeTab = -1;
	}

	public void SetActiveTab ( string n ) {
		SetActiveTab ( int.Parse ( n ) );
	}

	public void SetActiveTab ( int tab ) {
		activeTab = tab;
		
		for ( int i = 0; i < tabs.Count; i++ ) {
			if ( tabs[i].content ) {
				tabs[i].content.SetActive ( i == activeTab );
			}
		}
	}
	
	
	//////////////////
	// Mouse
	//////////////////
	override public void OnMouseDown () {
		if ( CheckMouseOver ( GetLeftRect () ) && activeTab > 0 ) {
			leftNavDown = true;

		} else if ( CheckMouseOver ( GetRightRect () ) && activeTab < tabs.Count - 1 ) {
			rightNavDown = true;

		} else {
			for ( int i = 0; i < tabs.Count; i++ ) {
				if ( CheckMouseOver ( GetTabRect ( i ) ) ) {
					SetActiveTab ( i );

					if ( target && !string.IsNullOrEmpty ( message ) ) {
						if ( passSelectedTab ) {
							target.SendMessage ( message, i );

						} else if ( !string.IsNullOrEmpty ( argument ) ) {
							target.SendMessage ( message, argument );

						} else {
							target.SendMessage ( message );

						}
					}
				}
			}
		}
	}

	override public void OnMouseUp () {
		if ( leftNavDown ) {
			SetActiveTab ( activeTab - 1 );
		
		} else if ( rightNavDown ) {
			SetActiveTab ( activeTab + 1 );

		}
		
		OnMouseCancel ();
	}

	override public void OnMouseCancel () {
		leftNavDown = false;
		rightNavDown = false;

		root.ReleaseWidget();
	}
	

	//////////////////
	// Update
	//////////////////
	override public void UpdateWidget () {
		isSelectable = true;
	
		// Mouse
		mouseRct = drawRct;

		// Update data
		overflow = GetTotalWidth () > drawRct.width;
		
		if ( activeTab >= tabs.Count && tabs.Count > 0 ) {
			SetActiveTab ( tabs.Count - 1 );
	 	}

		if ( !overflow ) {
			focusTab = 0;

		} else if ( activeTab < focusTab ) {
			focusTab--;
		
		} else if ( activeTab >= focusTab + maxTabs ) {
			focusTab++;

		}
	}

	
	//////////////////
	// Draw
	//////////////////
	override public void DrawSkin () {
		for ( int i = focusTab; i < focusTab + maxTabs; i++ ) {
			OGDrawHelper.DrawSlicedSprite ( GetTabRect(i), GetTabStyle(i), drawDepth, tint, clipTo );
		}
		
		if ( overflow ) {
			if ( activeTab > 0 ) {
				OGDrawHelper.DrawSlicedSprite ( GetLeftRect (), styles.thumb, drawDepth, tint, clipTo );
			}
			
			if ( activeTab < tabs.Count - 1 ) {
				OGDrawHelper.DrawSlicedSprite ( GetRightRect (), styles.thumb, drawDepth, tint, clipTo );
			}
		}
	}

	override public void DrawText () {
		for ( int i = focusTab; i < focusTab + maxTabs; i++ ) {
			OGDrawHelper.DrawLabel ( GetTabRect ( i ), tabs[i].title, GetTabStyle(i).text, drawDepth, tint, clipTo );
		}
		
		if ( overflow ) {
			if ( activeTab > 0 ) {
				OGDrawHelper.DrawLabel ( GetLeftRect (), leftArrow, styles.thumb.text, drawDepth, tint, clipTo );
			}

			if ( activeTab < tabs.Count - 1 ) {
				OGDrawHelper.DrawLabel ( GetRightRect (), rightArrow, styles.thumb.text, drawDepth, tint, clipTo );
			}
		}
		
	}
}
